Backpack/src/main/java/net/skyscanner/backpack/util/StateMachine.kt

```diff
@@ -0,0 +1,86 @@
+/**
+ * Backpack for Android - Skyscanner's Design System
+ *
+ * Copyright 2018-2021 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.skyscanner.backpack.util
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+interface StateMachine<State, Effect> {
+
+  val state: StateFlow<State>
+
+  val effects: SharedFlow<Effect>
+}
+
+internal interface MutableStateMachine<State, Effect> : StateMachine<State, Effect> {
+
+  interface CommitScope<Effect> {
+
+    fun emmit(effect: Effect)
+  }
+
+  fun commit(block: suspend CommitScope<Effect>.(State) -> State)
+}
+
+@Suppress("FunctionName")
+internal fun <State, Effect> MutableStateMachine(
+  scope: CoroutineScope,
+  initial: State,
+): MutableStateMachine<State, Effect> {
+
+  val state = MutableStateFlow(initial)
+  val effects = MutableSharedFlow<Effect>()
+  val mutex = Mutex()
+
+  return object : MutableStateMachine<State, Effect> {
+
+    override val effects: SharedFlow<Effect> = effects
+
+    override val state: StateFlow<State> = state
```

Comment:
what is the main reason for having a hot flow of data for the state in a Calendar? Do we have cases where it has to change in reaction to external interactions? My two cents is that a listener method or interface would cover the job here in a much simpler way.

Reply:
You may need to have a sync access to the state, as well as do some filtering operations as well. You may also want to know about some parameters changes or have some access to it.

This may be hidden beyond the listener, but I believe working with state is much better options. Thoughts?

Reply:
A callback would be helpful to actually be notified about _user_ updates, rather than updates the clients made - for example if the client updates some data on the calendar that is different to a user selecting a different range